File: src/main/java/org/kryptokrona/sdk/service/WalletService.java
// Copyright (c) 2022-2023, The Kryptokrona Developers
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification, are
// permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of
//    conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list
//    of conditions and the following disclaimer in the documentation and/or other
//    materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be
//    used to endorse or promote products derived from this software without specific
//    prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package org.kryptokrona.sdk.service;

import io.reactivex.rxjava3.core.Observable;
import org.kryptokrona.sdk.config.Config;
import org.kryptokrona.sdk.exception.node.NodeDeadException;
import org.kryptokrona.sdk.exception.wallet.WalletSubWalletNoPrimaryAddressException;
import org.kryptokrona.sdk.daemon.DaemonImpl;
import org.kryptokrona.sdk.util.CryptoUtils;
import org.kryptokrona.sdk.util.Metronome;
import org.kryptokrona.sdk.wallet.Address;
import org.kryptokrona.sdk.wallet.SubWallets;
import org.kryptokrona.sdk.wallet.Wallet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * WalletService.java
 *
 * @author Marcus Cvjeticanin (@mjovanc)
 */
public class WalletService {

	private final DaemonImpl daemon;

	private SubWallets subWallets;

	private WalletSynchronizerService walletSynchronizerService;

	private Metronome syncThread;

	private Metronome daemonUpdateThread;

	private Metronome lockedTransactionsCheckThread;

	private boolean started;

	private boolean synced;

	private boolean autoOptimize;

	private boolean shouldPerformAutoOptimize;

	private boolean currentlyOptimizing;

	private boolean currentlyTransacting;

	private static final Logger logger = LoggerFactory.getLogger(WalletService.class);

	public WalletService(
		DaemonImpl daemon
	) {
		this.daemon = daemon;
		this.started = false;
		this.autoOptimize = true;
		this.shouldPerformAutoOptimize = true;
		this.currentlyOptimizing = false;
		this.currentlyTransacting = false;
		this.setupMetronomes();
	}

	public SubWallets initializeSubWallets(
		Address address, long scanHeight, boolean newWallet, String privateViewKey, String privateSpendKey
	) {
		var subWallets = new SubWallets(address, scanHeight, newWallet, privateViewKey, privateSpendKey);
		var timestamp = 0L;

		if (newWallet) {
			timestamp = CryptoUtils.getCurrentTimestampAdjusted();
		}

		walletSynchronizerService = new WalletSynchronizerService(
				daemon, subWallets, timestamp, scanHeight, privateViewKey
		);

		/*if (!usingNativeCrypto()) {

		}*/

		return subWallets; // temporary return
	}

	//TODO: before we implement this we need to include the crypto library with JIT
	/*public boolean usingNativeCrypto() {
		logger.info("Checking if usage of native crypto.");

		return Crypto.isNative;
	}*/

	public void start() throws IOException {
		if (!started) {
			started = true;

			try {
				daemon.init();
				logger.info("Starting the wallet sync process.");

				var st = syncThread.start().publish();
				/*var dut = daemonUpdateThread.start().publish();
				var ltct = lockedTransactionsCheckThread.start().publish();*/

				st.subscribe(r -> sync(true).blockingSubscribe());
				/*dut.subscribe(r -> updateDaemonInfo().blockingSubscribe());
				ltct.subscribe(r -> checkLockedTransactions().blockingSubscribe());*/

				st.connect();
				/*dut.connect();
				ltct.connect();*/
			} catch (NodeDeadException e) {
				logger.error("%s", e);
			}
		}
	}

	public void stop() {
		started = false;
		// daemon.stop();

		syncThread.stop();
		daemonUpdateThread.stop();
		lockedTransactionsCheckThread.stop();

		logger.info("Stopping the wallet sync process.");
	}

	public Observable<Boolean> processBlocks(boolean sleep) throws NodeDeadException, InterruptedException {
		logger.info("Processing blocks...");
		var data = walletSynchronizerService.fetchBlocks().blockingSingle();

		if (data.size() != 0) {
			var shouldSleep = data.keySet().iterator().next();
			var blocks = data.values().iterator().next();

			if (blocks.size() == 0) {
				// shouldSleep should be a second condition below
				if (sleep && shouldSleep) {
					Thread.sleep(1000);
				}

				// we don't have any blocks, so we don't do any more processing
				return Observable.just(false);
			}

			for (var block : blocks) {
				logger.info("Processing block " + block.getBlockHeight());

				// forked chain, remove old data
				if (walletSynchronizerService.getHeight() >= block.getBlockHeight()) {
					logger.info("Removing forked transactions.");

					subWallets.removeForkedTransactions(block.getBlockHeight());
				}

				if (block.getBlockHeight() % 5000 == 0 && block.getBlockHeight() != 0) {
					subWallets.pruneSpentInputs(block.getBlockHeight() - 5000);
				}

				/* User can supply us a function to do the processing, possibly
				   utilizing native code for moar speed */
				// var processFunction = external

				var globalIndexes = new HashMap<String, List<Long>>();

				// for (var input : blockInputs)

				// var txData =

				// storeTxData(txData, block.getBlockHeight());

				// walletSynchronizer.dropBlock(block.getBlockHeight(), block.getBlockHash());

				logger.info("Finishing process block " + block.getBlockHeight());
			}
		}

		return Observable.just(true);
	}

	public Observable<Boolean> sync(boolean sleep) {
		try {
			return processBlocks(sleep);
		} catch (Exception e) {
			logger.error("Error processing blocks: " + e);
		}

		return Observable.just(false);
	}

	/**
	 * Update daemon status.
	 *
	 * @return Observable
	 */
	public Observable<Void> updateDaemonInfo() throws IOException, NodeDeadException {
		logger.info("Updating daemon info...");

		daemon.updateDaemonInfo();

		var walletHeight = walletSynchronizerService.getHeight();
		var networkHeight = daemon.getNetworkBlockCount();

		if (walletHeight >= networkHeight) {

			// synced with the network
			if (!synced) {
				synced = true;
			}

			if (shouldPerformAutoOptimize && autoOptimize) {
				try {
					performAutoOptimize();
				} catch (Exception e) {
					logger.error("Auto-optimize error.");
				}
			}
		}

		return Observable.empty();
	}

	/**
	 * Remove any transactions that have been cancelled.
	 *
	 * @return Observable
	 */
	public Observable<Void> checkLockedTransactions() {
		logger.info("Checking locked transactions...");

		var lockedTxHashes = subWallets.getLockedTransactionHashes();
		var cancelledTxs = walletSynchronizerService
			.findCancelledTransactions(lockedTxHashes)
			.blockingSingle();

		for (var cancelledTx : cancelledTxs) {
			subWallets.removeCancelledTransaction(cancelledTx);
		}

		return Observable.empty();
	}

	public void setupMetronomes() {
		syncThread = new Metronome(Config.SYNC_THREAD_INTERVAL);
		daemonUpdateThread = new Metronome(Config.DAEMON_UPDATE_INTERVAL);
		lockedTransactionsCheckThread = new Metronome(Config.LOCKED_TRANSACTIONS_CHECK_INTERVAL);
	}

	/**
	 * Creates a new wallet instance with a random key pair.
	 *
	 * @return Wallet
	 */
	public Wallet createWallet() {
		logger.info("Creating new wallet.");

		var newWallet = true;
		var scanHeight = 0;
		var address = Address.fromEntropy("", "").blockingSingle();

		subWallets = initializeSubWallets(
			address, scanHeight, newWallet, address.getViewKeys().getPrivateKey(),
			address.getSpendKeys().getPrivateKey());

		//TODO: finish implementation here

		createSubWallets();
		return new Wallet();
	}

	/**
	 * Creates SubWallets for a wallet.
	 */
	public void createSubWallets() {
		logger.info("Creating subwallets.");
		// await keys.address(), scanHeight, newWallet,
		//            keys.view.privateKey, keys.spend.privateKey,
		// this.subWallets = new SubWallets();

		//TODO: finish implementation here
	}

	/**
	 * Gets the private spend and private view for the primary address.
	 * The primary address is the first created wallet in the container.
	 *
	 * @return Observable
	 */
	public Observable<HashMap<String, String>> getPrimaryAddressPrivateKeys()
		throws WalletSubWalletNoPrimaryAddressException
	{
		logger.info("Getting primary address private keys.");
		var keys = new HashMap<String, String>();
		keys.put(subWallets.getPrimaryPrivateSpendKey(), subWallets.getPrivateViewKey());

		return Observable.just(keys);
	}

	/**
	 * Get the primary address mnemonic seed. If the primary address isn't
	 * a deterministic wallet, it will return a WalletError.
	 *
	 * @return Observable
	 */
	public Observable<ArrayList<String>> getMnemonicSeed() throws WalletSubWalletNoPrimaryAddressException {
		logger.info("Getting mnemonic seed.");

		var mnemonic = getMnemonicSeedForAddress(subWallets.getPrimaryAddress()).blockingSingle();

		return Observable.just(mnemonic);
	}

	/**
	 * Get the mnemonic seed for the specified address. If the specified address
	 * is invalid or the address isn't a deterministic wallet, it will return
	 * a WalletError.
	 *
	 * @param address A valid address that exists in this container
	 * @return Observable
	 */
	public Observable<ArrayList<String>> getMnemonicSeedForAddress(String address) {
		logger.info("Getting mnemonic seed for address");

		var mnemonicSeed = new ArrayList<String>();

		//TODO more implementation here

		return Observable.just(mnemonicSeed);
	}

	/**
	 * Since we're going to use optimize() with auto optimizing, and auto
	 * optimizing is enabled by default, we have to ensure we only optimize
	 * a single wallet at once. Otherwise, we'll end up with everyones balance
	 * in the primary wallet.
	 *
	 * @return Observable
	 */
	public Observable<Map<List<String>, Long>> optimizeAddress(String address) {
		return Observable.empty();
	}

	/**
	 * Optimizes your wallet as much as possible. It will optimize every single
	 * subwallet correctly, if you have multiple subwallets. Note that this
	 * method does not wait for the funds to return to your wallet before
	 * returning, so, it is likely balances will remain locked.
	 *
	 * Note that if you want to alert the user in real time of the hashes or
	 * number of transactions sent, you can subscribe to the `createdfusiontx`
	 * event. This will be fired every time a fusion transaction is sent.
	 *
	 * You may also want to consider manually creating individual transactions
	 * if you want more control over the process. See [[sendFusionTransactionBasic]].
	 *
	 * This method may take a *very long time* if your wallet is not optimized
	 * at all. It is suggested to not block the UI/mainloop of your program
	 * when using this method.
	 */
	public Observable<Map<ArrayList<String>, Long>> optimize() {
		logger.info("Method optimize called.");

		var numTxsSent = 0L;
		var hashes = new ArrayList<String>();

		for (var address : subWallets.getAddresses()) {
			var data = optimizeAddress(address).blockingSingle();
			numTxsSent += data.values().iterator().next();
			hashes.addAll(data.keySet().iterator().next());
		}

		return Observable.just(Map.of(hashes, numTxsSent));
	}

	public Observable<Void> performAutoOptimize() {
		shouldPerformAutoOptimize = false;

		// already optimizing, don't optimize again
		if (!currentlyOptimizing) {
			currentlyOptimizing = true;

			if (!currentlyTransacting) {
				logger.info("Performing auto optimization.");

				optimize();

				logger.info("Auto optimization complete.");
			}
		}

		// we're done with optimizing
		currentlyOptimizing = false;

		return Observable.empty();
	}

}

File: kryptokrona-core/src/main/kotlin/org/kryptokrona/sdk/core/service/WalletService.kt
// Copyright (c) 2022-2023, The Kryptokrona Developers
//
// Written by Marcus Cvjeticanin
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification, are
// permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of
//    conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list
//    of conditions and the following disclaimer in the documentation and/or other
//    materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be
//    used to endorse or promote products derived from this software without specific
//    prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package org.kryptokrona.sdk.core.service

import kotlinx.coroutines.*
import org.kryptokrona.sdk.crypto.Crypto
import org.kryptokrona.sdk.crypto.getKeyImageFromOutput
import org.kryptokrona.sdk.crypto.model.TransactionInput
import org.kryptokrona.sdk.crypto.util.convertHexToBytes
import org.kryptokrona.sdk.http.client.BlockClient
import org.kryptokrona.sdk.http.client.WalletClient
import org.kryptokrona.sdk.http.model.request.block.BlockDetailsByHeightRequest
import org.kryptokrona.sdk.http.model.request.wallet.WalletSyncDataRequest
import org.kryptokrona.sdk.http.model.response.node.Info
import org.kryptokrona.sdk.http.model.response.walletsyncdata.Block
import org.kryptokrona.sdk.http.model.response.walletsyncdata.Transaction
import org.kryptokrona.sdk.http.model.response.walletsyncdata.WalletSyncData
import org.kryptokrona.sdk.util.config.Config
import org.kryptokrona.sdk.util.model.node.Node
import org.slf4j.LoggerFactory

/**
 * WalletService class.
 *
 * @author Marcus Cvjeticanin
 * @since 0.1.1
 * @param node The node that the wallet service is connected to.
 */
class WalletService(node: Node) {

    private val logger = LoggerFactory.getLogger("WalletService")

    /**
     * The crypto external library from C code.
     */
    private val crypto = Crypto()

    /**
     * The wallet client, used for sending requests to a node.
     */
    private val walletClient = WalletClient(node)

    /**
     * The block client, used for sending requests to a node.
     */
    private val blockClient = BlockClient(node)

    /**
     * The node service, used for getting information of the node.
     */
    private val nodeService = NodeService(node)

    /**
     * The sync job for the wallet sync process.
     */
    private var syncJob: Job = Job()

    /**
     * The node info we get from node service.
     */
    private var nodeInfo: Info? = null

    /**
     * If we don't set the start height, we will start from the genesis block.
     */
    private var startHeight: Long = 0

    /**
     * The current height of the wallet. If we set a start height, we will start from that height.
     */
    private var walletHeight: Long = 0

    /**
     * Checkpoints for the wallet sync process.
     */
    private var checkpoints: MutableList<String> = mutableListOf()

    /**
     * Stored blocks for later processing
     */
    private var storedBlocks = mutableListOf<Block>()

    fun getNodeInfo() = nodeInfo

    fun getStoredBlocks() = storedBlocks

    fun setStartHeight(height: Long) {
        startHeight = height
    }

    /**
     * Starts the sync process.
     *
     * @author Marcus Cvjeticanin
     * @since 0.2.0
     */
    suspend fun startSync() = coroutineScope {
        logger.info("Starting sync process...")
        walletHeight = startHeight

        // get the first hash for the checkpoints
        val blockDetails = blockClient.getBlockDetailsByHeight(BlockDetailsByHeightRequest(startHeight))
        blockDetails?.block?.hash?.let { checkpoints.plusAssign(it) }

        var lastCheckPoint = listOf(checkpoints.last())
        val requestData = WalletSyncDataRequest(blockHashCheckpoints = lastCheckPoint)
        val walletSyncData = getSyncData(requestData)
        walletSyncData.let { wsd ->
            if (wsd != null) {
                checkpoints.plusAssign(wsd.items.last().blockHash)
                walletHeight += wsd.items.size.toLong()
            }
        }

        syncJob = launch {
            launch(Dispatchers.IO) {
                while(isActive) {
                    logger.debug("Syncing blocks...")

                    nodeInfo?.height?.let {
                        if (walletHeight < it) {
                            lastCheckPoint = listOf(checkpoints.last())
                            val data = WalletSyncDataRequest(blockHashCheckpoints = lastCheckPoint)
                            val syncData = getSyncData(data)

                            syncData?.let { sd ->
                                if (sd.items.isNotEmpty()) {
                                    checkpoints.plusAssign(sd.items.last().blockHash)
                                    checkpoints = checkpoints.distinct().toMutableList() // removing duplicates
                                    walletHeight += sd.items.size.toLong()

                                    // add new blocks to stored blocks
                                    sd.items.forEach { block -> storedBlocks.add(block) }
                                    logger.info("Fetched ${sd.items.size} block(s)")
                                }
                            }

                            processBlocks()

                            logger.info("Wallet height: $walletHeight")
                        }
                    }

                    delay(Config.SYNC_THREAD_INTERVAL)
                }
            }

            launch(Dispatchers.IO) {
                while(isActive) {
                    nodeInfo = nodeService.getNodeInfo()
                    logger.info("Node height: ${nodeInfo?.height}")
                    delay(Config.NODE_UPDATE_INTERVAL)
                }
            }
        }

        syncJob.children.forEach { it.join() }
    }

    /**
     * Stops the sync process.
     *
     * @author Marcus Cvjeticanin
     * @since 0.2.0
     */
    suspend fun stopSync() = coroutineScope {
        syncJob.children.forEach { it.cancel() }

        logger.info("Stopping sync process...")
    }

    /**
     * Gets the wallet sync data.
     *
     * @author Marcus Cvjeticanin
     * @since 0.2.0
     * @return The wallet sync data.
     */
    private suspend fun getSyncData(walletSyncDataRequest: WalletSyncDataRequest): WalletSyncData? {
        logger.debug("Getting wallet sync data...")
        return walletClient.getWalletSyncData(walletSyncDataRequest)
    }

    /**
     * Processes the stored blocks and checks the transactions
     * for outputs that belong to the wallet. Removes the processed
     * blocks from the storedBlocks list when done.
     *
     * @author Marcus Cvjeticanin
     * @since 0.2.0
     */
    private fun processBlocks() {
        logger.info("Processing blocks...")

        val blocksToRemove = mutableListOf<Block>()

        storedBlocks.forEach { block ->
            block.transactions.forEach { transaction ->
                checkTransactionOutputs(transaction, block.blockHeight)
            }

            // update wallet scan height
            walletHeight = block.blockHeight

            // remove the checked block from storedBlocks
            blocksToRemove.add(block)
        }

        storedBlocks.removeAll(blocksToRemove)
    }

    /**
     * Checks the transaction outputs for outputs that belong to the wallet
     * and adds them to the wallet.
     *
     * @author Marcus Cvjeticanin
     * @since 0.2.0
     * @param transaction The transaction to check.
     * @param blockHeight The block height of the transaction.
     */
    private fun checkTransactionOutputs(transaction: Transaction, blockHeight: Long) {
        val publicSpendKey = "cde60afedba1e88a9c7e8b28cc038ee018d5a24a1a239cdcb8d32506a594f3cb"
        val privateViewKey = "8f066e33d45a0205b772f47b5a5d66f5b5e08fc329c45fc5f2a15a998ad0d4b4"

        val pubSpend = convertHexToBytes(publicSpendKey)
        val privView = convertHexToBytes(privateViewKey)
        val txPubKey = convertHexToBytes(transaction.txPublicKey)

        // val inputs = mutableListOf<TransactionInput>()

        val derivation = ByteArray(32)
        crypto.generateKeyDerivation(txPubKey, privView, derivation)

        transaction.outputs.forEachIndexed { index, output ->
            val key = output.key
            val derivedKey = convertHexToBytes(key)
            val base = ByteArray(32)

            crypto.underivePublicKey(derivation, index.toLong(), derivedKey, base)

            // if the derived spend key does not match, the output key is not designated to us.
            // continue checking other outputs.
            if (!pubSpend.contentEquals(derivedKey)) {
                return@forEachIndexed
            }

            // this transaction contains outputs that belong to us.
            // create the key image and transaction input and save it
            val keyImage = getKeyImageFromOutput(derivation, index.toLong(), pubSpend)

            // this is not spent yet, we just got it :)
            val spendHeight = 0

            // construct our transaction input, there may be more inputs from this transactions
            val txInput = TransactionInput(
                amount = output.amount,
                blockHeight = blockHeight,
                keyImage = keyImage,
                txPubKey = transaction.txPublicKey,
                key = key,
                privateEphemeral = keyImage.privateSpendKey,
                txHash = transaction.hash,
                txIndex = index.toLong(),
                spendHeight = 0,
                unlockTime = transaction.unlockTime,
                globalIndex = 0
            )

            logger.info("Transaction found with hash: " + txInput.txHash)

            // TODO we add this input to its wallet
            // inputs.add(txInput)
        }
    }
}
